src/core/provider.cc

```diff
@@ -787,4 +809,148 @@ InferResponseProvider::FinalizeResponse(const InferenceBackend& is)
+
+  provider->request_header_ = request_header;
+
+  RETURN_IF_ERROR(provider->NormalizeRequestHeader(is));
+
+  for (const auto& io : provider->request_header_.input()) {
+    auto it = input_blocks.find(io.name());
+    if (it == input_blocks.end()) {
+      return Status(
+          RequestStatusCode::INVALID_ARG,
+          "input '" + io.name() + "' is specified in request header but" +
+              " not found in memory block mapping for model '" +
+              provider->model_name_ + "'");
+    }
+    if (io.batch_byte_size() != it->second->ByteSize()) {
+      return Status(
+          RequestStatusCode::INVALID_ARG,
+          "unexpected size " + std::to_string(it->second->ByteSize()) +
+              " for input '" + io.name() + "', expecting " +
+              std::to_string(io.batch_byte_size()) + " for model '" +
+              provider->model_name_ + "'");
+    }
+    provider->input_blocks_[io.name()] = std::make_pair(it->second, false);
+  }
+
+  return Status::Success;
+}
+
+Status
+SystemMemoryInferRequestProvider::GetNextInputContent(
+    const std::string& name, const void** content, size_t* content_byte_size,
+    bool force_contiguous)
+{
+  if (*content_byte_size == 0) {
+    *content = nullptr;
+    return Status::Success;
+  }
+
+  if (!GetInputOverrideContent(name, content, content_byte_size)) {
+    const auto& pr = input_blocks_.find(name);
+    if (pr == input_blocks_.end()) {
+      return Status(
+          RequestStatusCode::INTERNAL, "unexpected input '" + name + "'");
+    }
+
+    auto& input_content = pr->second;
+
+    if (input_content.second) {
+      *content = nullptr;
+      *content_byte_size = 0;
+    } else {
+      *content = input_content.first->Buffer();
+      *content_byte_size = input_content.first->ByteSize();
+    }
+
+    input_content.second = true;
+  }
+
+  return Status::Success;
+}
+
+Status
+SystemMemoryInferResponseProvider::Create(
+    const InferenceBackend& is, const InferRequestHeader& request_header,
+    std::shared_ptr<SystemMemoryInferResponseProvider>* infer_provider)
+{
+  auto provider = new SystemMemoryInferResponseProvider(request_header);
+  infer_provider->reset(provider);
+  return Status::Success;
+}
+
+const InferResponseHeader&
+SystemMemoryInferResponseProvider::ResponseHeader() const
+{
+  return response_header_;
+}
+
+InferResponseHeader*
+SystemMemoryInferResponseProvider::MutableResponseHeader()
+{
+  return &response_header_;
+}
+
+Status
+SystemMemoryInferResponseProvider::GetOutputBuffer(
+    const std::string& name, void** content, size_t content_byte_size,
+    const std::vector<int64_t>& content_shape)
+{
+  *content = nullptr;
+
+  Output* output;
+  RETURN_IF_ERROR(CheckAndSetIfBufferedOutput(
+      name, content, content_byte_size, content_shape, &output));
+
+  // Always write output tensor to an output block no matter
+  // if output has cls field defined
+  if (output_blocks_.find(name) != output_blocks_.end()) {
```

Comment:
Performance. Get the iterator to output_blocks_[name] once and then use it in all places below.